### ISSUE TITLE
6581 Filter Volumes

### ISSUE THREAD

**Issue by acrouzet** [libsidplayfp/libsidplayfp#139]
I've noticed that the volume of channels under the 6581 filter can sound a bit louder in the emulation than on, at least some, real 6581s. This seems to be true regardless of filter response parameters.

This difference in volume between real 6581s and emulation seems to depend on the filter mode selected. Low-pass doesn't sound much different, while band pass, high pass, and combined modes have the most difference. It may also depend on the amplitude of the signal being filtered, as some tunes with the same filter characteristics don't sound as off to me.

I'm unsure whether the relative volume of the filter modes varies much from chip to chip, as I don't have a large enough sample size (or a good enough ear) to know for sure.

But for an example to test, compare the volume of the arps in this tune, which are under a combined low, band, and high pass filter: /MUSICIANS/B/Blues_Muz/Gallefoss_Glenn/Acid_Jazz.sid (Perhaps look for tunes where filtered channels are in similarly prominent positions as well.)

**Comment by reFX-Mike** [comment]
Listen to Artillery from Geir Tjelta. Uses all filter modes at the same time for the main melody. Sounds perfectly fine and balanced. It’s Acid Jazz that stands out as a negative example, and I think it’s the song itself, not the emulation that is bad.

I’ve listened to literally hundreds of tunes over and over then last two years. The emulation is fine.

**Comment by reFX-Mike** [comment]
I could imagine that this is again chip-dependent. So some chips will have a louder (and also more distorted) filter, and other chips are a bit more tame.

The current hardcoded values (for filter-cutoff offset, filter-gain, etc.) are best suited for the 6581 that Martin Galway had. His tunes sound perfect, but for other 6581 tunes, different cutoff-offsets are needed, and perhaps the filter-gain needs to be adjustable as well.

I've started work on `chip-profiles` which is a list of 6581 cutoff offset and digi-volume settings, mapped to folders in the HVSC. That way I can control filter adjustments for Rob Hubbard and have them different than Martin Galway.

I will add filter-gain as an adjustable parameter and play with it to see if we can get more pleasant results for certain authors. Glenn Gallefoss probably needs a lower filter gain to sound more balanced.

**Comment by reFX-Mike** [comment]
After some experimentation, I found that add this line:

`Vsum = ( Vsum * 3276 ) >> 12;`

before
https://github.com/libsidplayfp/libsidplayfp/blob/9cc3421737837558d360105dcc9ace9eb7c500f9/src/builders/residfp-builder/residfp/Filter.h#L212

makes pretty much **ALL** 6581 tunes more balanced. Acid Jazz benefits the most, but even tunes I thought were pretty good before, sound even better now. Even Martin Galway tunes.

What this one line does is multiply the summed signal, that goes into the filter, by 80%. I've used only 12 bits, so the multiply wouldn't cause an overflow for extreme input. The multiplier (3276 in this test) can obviously be changed to an int variable that can go from 0 (no filter output at all) up to 4096 = old volume = too loud.

For whatever reason this one little change raises the digi-volume quite a bit. So the DC offset for the 6581 has to be adjusted to avoid distortion and restore balance in that regard.

**Comment by reFX-Mike** [comment]
I've changed my code slightly to do this now:

```
constexpr auto	filterGainOut = int ( 0.85f * 256 );

if ( lp )	Vmix += ( Vlp * filterGainOut ) >> 8;
if ( bp )	Vmix += ( Vbp * filterGainOut ) >> 8;
if ( hp )	Vmix += ( Vhp * filterGainOut ) >> 8;
```

And I've removed my previous line that scaled the filter-input

This new block of code scales the filter output to 85% and with all the tunes I've tested, it sounds way more balanced. This is only true for the 6581 filter, so my branch still uses two different `clock` functions. My changes are only applied to the Filter6581.h `clock` function. The Filter8580.h `clock` function is untouched.

**Comment by drfiemost** [comment]
But this doesn't match with the actual circuit, where all the lines enter the mixer at the same level.
It's more likely related to the integrators, maybe the missing slope factor which I can't get right. Or maybe the caps values off from the spec.

**Comment by drfiemost** [comment]
Yes, it's analog. However, while the absolute resistance may vary from chip to chip, the resistor ratio should be quite stable.

For those interested the schematic is here:
https://github.com/libsidplayfp/SID_schematics/blob/master/schematics/page13.png

**Comment by reFX-Mike** [comment]
I don't even think this is a chip-to-chip issue. All 6581 tunes sounded more balanced and pleasant when reducing the filter output to 80-85%. Even for tunes whose chip profiles were extremely different from each other. It seems to be a constant.

I was unsuccessful in modifying the filter code, as it's all Greek to me. The modeling of analog components in C/C++ code seems very complex and outside my realm of competency.

That's why I originally did the output scaling. It's a hacky workaround, but that's why I'm used to from the video games and even the pro-audio world: fake it until it sounds/looks good.

Hopefully, you can come up with a more robust solution.

**Comment by drfiemost** [comment]
I've been looking again at the schematics ([this one](https://raw.githubusercontent.com/libsidplayfp/SID_schematics/master/schematics/page12.png) is a bit easier to read) and in fact the resistance for the filter lines entering the mixer is slightly bigger than that of the voices.
The estimated values are 251.54K vs 233.67K while the opamp feedback resistance is 330.05K. So it looks more like an ~93% attenuation.

i.e. 330/251 is the n=8/6 value we pass to the opamp model here:
https://github.com/libsidplayfp/libsidplayfp/blob/18882a4e4dcea18004e81175db9e95daab7955ea/src/builders/residfp-builder/residfp/FilterModelConfig6581.cpp#L165
then a further reduction of 233/251~=0.93 for the filter inputs.

**Comment by drfiemost** [comment]
Checked the vectorized image and it confirm the findings. The two circled resistors (the yellowish lines) are slightly different, the one on the left being a bit longer, meaning more resistance.
![mixer_res_6581](https://github.com/user-attachments/assets/4acf9dfd-cb49-4c9b-9f31-ad48bccb760f)
It seems more accidental than a design choice.

Fix is available in #144, based on reFX-Mike suggestion, let me know if it sounds good.

**Comment by reFX-Mike** [comment]
I like it better now. In my branch, I went as low as 0.85 because, for most songs, it sounds even more pleasant, but then a few digi tunes sound a bit noisier (e.g., Kapla Caves) than with 0.93, but that's a price I'm willing to pay for better overall non-filter/filter balance.

With 0.93, there is virtually no difference for Kapla Caves. Sounds super clean.
